### PR TITLE
Remove filterRight.

### PR DESCRIPTION
**Pull request by iclanzan**

### CODE REVIEW

File: src/animationLoop.js [hunk 1]
```diff
@@ -74,16 +68,31 @@ const prototype = {
...
-      renderSubscriber,
-      1 + animationLoop.accumulatedTime / timeStep
-    );
+    const alpha = 1 + animationLoop.accumulatedTime / timeStep;
+    let index = 0;
+    let nextState;
+    let subscriber;
+
+    while (index < state.length) {
+      subscriber = state[index];
```
Comment (chenglou) [inline src/animationLoop.js]: If you inline the `let` here babel will transform it into a regular `var`. Probably _slightly_ more readable.

**Comment by chenglou** [comment]
As discussed, this is indeed less readable. But could allocate a lot less if you do that opt where you check if all subscribers are active right?

**Comment by iclanzan** [comment]
Absolutely. I think it’s a nice optimization.

**Comment by chenglou** [comment]
So are we gonna do the optimization or leave it for later?

**Comment by iclanzan** [comment]
What do you mean? The optimization is right there. Line 84 only allocates a new array if a subscriber is not active.

**Comment by iclanzan** [comment]
Looking at it I realize it’s flawed. It would miss adding active subscribers that are positioned before the first inactive one.

**Comment by chenglou** [comment]
Yeah I think ironically, I got confused but this code and thought you were doing something else.

**Comment by chenglou** [comment]
So, do you still think it's worth doing this or use the `filterRight`?

**Comment by iclanzan** [comment]
Ok, fixed it.

**Comment by iclanzan** [comment]
We should definitely **not** use `filterRight` since it’s `O(n^2)`.

**Comment by chenglou** [comment]
Sorry meant to say #78

**Comment by iclanzan** [comment]
It’s up to you. That one is more readable and easier to test. This one is more optimized.

File: src/animationLoop.js [hunk 2]
```diff
@@ -32,6 +25,13 @@ const prototype = {
         subscriber.prevValue = value;
         subscriber.value = subscriber.step(timeStep, value);
       }
+
+      // Can’t use else here because subscriber.step() might cause
+      // subscriber to become inactive.
+      if (!subscriber.active) {
+        // Mark as dirty to perform filtering in the render step.
+        animationLoop.dirty = true;
```
Comment (nkbt) [inline src/animationLoop.js]: Can't see where `dirty` is reset. Is it a bug or feature?

File: src/animationLoop.js [hunk 3]
```diff
@@ -74,16 +75,30 @@ const prototype = {
...
-      1 + animationLoop.accumulatedTime / timeStep
-    );
+    const alpha = 1 + animationLoop.accumulatedTime / timeStep;
+    let index = 0;
+    let nextState;
+
+    // If there are any inactive subscribers (loop is dirty)
+    // we need to filter state.
+    if (animationLoop.dirty) {
+      nextState = animationLoop.state = [];
```
Comment (nkbt) [inline src/animationLoop.js]: I would avoid double assignments.

File: src/animationLoop.js [hunk 4]
```diff
@@ -74,16 +75,30 @@ const prototype = {
...
+    }
+
+    while (index < state.length) {
```
Comment (nkbt) [inline src/animationLoop.js]: thous should perfectly work with old good `for` loop instead.

Comment (iclanzan) [inline src/animationLoop.js]: True, I just have a preference for `while`.

File: src/animationLoop.js (hunk 3 above)
Comment (iclanzan) [inline src/animationLoop.js]: Are you saying you would rather have the following?
```
nextState = [];
animationLoop.state = nextState;
```

File: src/animationLoop.js (hunk 4 above)
Comment (nkbt) [inline src/animationLoop.js]: Yeah, you've implemented `for` loop with incremented guard with `while` :).

File: src/animationLoop.js (hunk 2 above)
Comment (iclanzan) [inline src/animationLoop.js]: Bug. Good catch!

**Comment by chenglou** [comment]
Keeping this open as a task for when we get enough test coverage to replace #78. We'll get there eventually =)

File: src/animationLoop.js (hunk 3 above)
Comment (nkbt) [inline src/animationLoop.js]:
```js
nextState = [];
animationLoop.state = [];
```
If you want to actually keep mutable reference, then yes
```js
nextState = [];
animationLoop.state = nextState;
```

**Comment by chenglou** [comment]
Diverged, closing but will use as reference.

**Pull request closed without merging**